### ISSUE TITLE
Clarification about mixing_rate

### ISSUE THREAD

**Issue by danyaljj** [google-research/text-to-text-transfer-transformer#200]
I am trying to make sure I understand definition of `default_rate`: https://github.com/google-research/text-to-text-transfer-transformer/blob/master/t5/data/utils.py#L875-L879

Based on the definition "relative numbers of examples to use from their associated tasks" if I mixture defined as follows: 

```python 
t5.data.MixtureRegistry.add(
       f"my_mixture", [task1, task2], default_rate=1.0
)
```
- Question1: this mixture definition means that we're gonna use 100% of both task1 and task2 for my training and evaluation, regardless of their absolute size; right? 

- Question2: I think the notion is `default_rate` is *independent* of how we sample instances when preparing our batches. To be concrete, suppose the training set of task1 is 7 times larger than task2 training set. That means that each batch of size 8 would, on average, have 7 training examples from task1 and 1 example from task 2. Is that correct?

**Comment by sharannarang** [comment]
Re: q1, yes that is correct. All the examples in both tasks will be used for training and eval.

Re: q2, With `default_rate=1.0`, each batch on average would have equal number of examples from both task1 and task2. If you're looking for rate proportional to number of examples in the training set, you could use something like this:

```
def _rate_num_input_examples(task):
  if "train" in task.splits:
    return float(task.num_input_examples("train"))
  elif "validation" in task.splits:
    return float(task.num_input_examples("validation"))
  else:
    raise ValueError("Task %s does not have a train or validation split." % (
        task.name))

t5.data.MixtureRegistry.add(
       f"my_mixture", [task1, task2], default_rate=_rate_num_input_examples
)
```

Please let me know if you have any further questions.

**Comment by nikett** [comment]
If the default rate is set to 0.1 on a singleton mixture, does it mean that 10% (random) data would be used for training and eval?

t5.data.MixtureRegistry.add(
       f"my_mixture", [task1], default_rate=0.1
)

**Comment by craffel** [comment]
No, a mixture with one task is equivalent to just using the task. The rates are normalized to sum to 1, effectively. If you only want to use 10% of a dataset, you can use the splitting and slicing API from TFDS when constructing your task https://github.com/google-research/text-to-text-transfer-transformer/blob/master/t5/data/utils.py#L803

**Comment by nikett** [comment]
Thanks for the clarification, Colin!

**Comment by nikett** [comment]
Assuming that we are not performing any packing, will the following be correct?

Suppose a mixture comprises k tasks `t1... tk` : 

case 1: if the default rate is a constant between 0.1 - 1.0:
then, ```number of steps = max(t1_num_examples, t2_num_examples, ... tk_num_examples) *k / batchsize```
If there is a skew in the number of examples between t1.. tk then, many examples from the task with fewest number of examples will be repeated across batches.

case 2: by using the _rate_num_input_examples function [earlier in this issue](https://github.com/google-research/text-to-text-transfer-transformer/issues/200#issuecomment-621626526):
```number of steps = sum(t1_num_examples, t2_num_examples, ... tk_num_examples) / batchsize```
Examples probably will not be repeated, and the model might pay more attention to the task with most examples.

**Comment by craffel** [comment]
Hey Niket, 

> `number of steps = max(t1_num_examples, t2_num_examples, ... tk_num_examples) *k / batchsize`

I'm not sure what you're trying to calculate here. The number of steps you train a model can be set independently of all of the quantities on the right of your equals sign.

> If there is a skew in the number of examples between t1.. tk then, many examples from the task with fewest number of examples will be repeated across batches.

Yes, this is true.

> Examples probably will not be repeated, and the model might pay more attention to the task with most examples.

Yes, all tasks will complete an epoch after the same number of training steps. The model will see the most data from the tasks with the largest number of examples.

**Comment by nikett** [comment]
Thanks for confirming that Collin. I meant "number of gradient steps in one epoch". It is clear now
